### PR TITLE
kubeadm: add separate page for configuring / migrating cgroup driver

### PR DESCRIPTION
**Pull request by neolit123**
- Add the new page "Configuring a cgroup driver"
The page includes details on how to configure the cgroup driver
using kubeadm. It also includes a migration guide for users to
move to the "systemd" driver.
- Link to the new page from install-kubeadm.md and container-runtimes.md

xref
https://github.com/kubernetes/kubeadm/issues/2376
https://github.com/kubernetes/kubernetes/pull/99471
xref https://github.com/kubernetes/website/issues/22916

note: this is targeting the 1.21 release

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:37]: this is new to kubeadm 1.21

**Comment by neolit123** [comment]
new page preview:
https://deploy-preview-26786--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:25]: What about moving this sentence and the beginning of the paragraph so first we are telling the user they should use systemd and why, and then we explain how.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (fabriziopandini) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:37]: Let's add "starting for v1.21" and "for older releases, the user should executed following steps manually"

Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:37]: i think the point of the versioned docs (this PR is targeting dev-1.21) is that all details apply to the target release - i.e. 1.21.

maybe we can use some sort of version tag for this.
will defer to sig-docs for this question.

**Review by onlydole — APPROVED** [review]
LGTM! Thank you, @neolit123!

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (pacoxu) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:114]: Sometimes, we need to clean up the container runtime dir.
Not sure if this is needed.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (pacoxu) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:112]: Should we link to https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/ for some details.
Most likely, user may use --force for application pods.

**Comment by pacoxu** [comment]
LGTM

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (binchenX) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:60]:
```suggestion
The `KubeletConfiguration` will be stored in a [ConfigMap](docs/concepts/configuration/configmap)
```

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (binchenX) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:61]:
```suggestion
object under the `kube-system` namespace after the control plane is up.
```

**Review by binchenX — COMMENTED** [review]
some nit.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:52]: i agree with the point that the kubelet if not run with systemd does not need to use the systemd driver.

but `container-runtimes.md` is really about the kubelet package (DEB/RPM) that k8s distributes and that package installs the kubelet as a systemd service, which means it should use the systemd driver.

also all the container runtimes on the page include steps to setup using the systemd driver.

Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:52]: > So I would suggest to keep existing content (which is still true and useful) and add a section called migration to systemd cgroup driver using kubeadm to make it explicit.

the change on this page is removing a `{{< caution >}}` blob about cgroup driver migration and replacing it with a guide for that. what other existing content do you mean?

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:61]: IMO this feels more confusing than clarifying.
to be precise it is stored after the first API server's health reports 200.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:60]: this shifts the paragraph from present tense to future tense.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:112]: i can include the link.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:114]: from my tests this was not needed.
we can update the guide after this PR merges if we find gaps.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:25]: makes sense.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (binchenX) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:61]: yeah, your description is more accurate. We can leave it as your original wording.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (binchenX) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:60]: I was trying to express the future tense.. since the "store" happens after one of the api server is running.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (binchenX) [reply content/en/docs/setup/production-environment/container-runtimes.md:52]: >what other existing content do you mean?
I mean the existing `{{<caution >}}` blob.

>also all the container runtimes on the page include steps to setup using the systemd driver.

I concerned more about pointing the instruction to a `kubeadm` page than assuming `systemd` is used. In reality, as far as I know, most setup do one systemd; but not all of them are using kubeadm for the bootstrap.

Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:52]: ok, i see what you mean.
i can preserve the caution but rephrase it a little to say that it is a "disruptive operation" instead of saying "not recommended", because it's actually recommended.

also rephrase the new section to explain that this is a kubeadm guide:

```
### Migrating to the `systemd` driver in kubeadm clusters

Follow the [Migration guide](/tasks/administer-cluster/kubeadm/configure-cgroup-driver)
if you wish to migrate to the `systemd` cgroup driver in existing kubeadm clusters.
```

Comment (binchenX) [reply content/en/docs/setup/production-environment/container-runtimes.md:52]: perfect. thanks.
and maybe "%s/kubeadm clusters/kubeadm managed clusters/g"

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:77]:
```suggestion
See the below section on "Modify the kubelet ConfigMap" for details on
```
nit

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:87]:
```suggestion
steps outlined below.
```

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:91]:
```suggestion
that use the `systemd` driver. This requires executing only the first step below
```

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:37]: added: `{{< feature-state for_k8s_version="v1.21" state="stable" >}}` inside the note.
not sure if that is the preferred way.

**Comment by neolit123** [comment]
updated the PR.

**Review by binchenX — APPROVED** [review]
lgtm

**Comment by rdxmb** [comment]
can we please remove
> fixes #22916

because of https://github.com/kubernetes/website/issues/22916#issuecomment-805645549

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (rdxmb) [inline content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:52]: https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc uses 

```
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
nodeRegistration:
  kubeletExtraArgs:
    cgroup-driver: "cgroupfs"
```

to define the cgroupfs.

Confusing...

Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:52]: don't use kubeletExtraArgs, because the cgroup-driver flag is deprecated.
this new page has the right config example.

**Comment by neolit123** [comment]
> can we please remove
> 
> > fixes #22916
> 
> because of [#22916 (comment)](https://github.com/kubernetes/website/issues/22916#issuecomment-805645549)

yes, we can.

File: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:52]: there are also the support forums and #kubeadm on the k8s slack for users that need help:
https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

i will send a PR to remove the kubeletExtraArgs examples in https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc

Comment (rdxmb) [reply content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md:52]: > there are also the support forums and #kubeadm on the k8s slack for users that need help:
https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

I know. But the docs should be the first place to search, shouldn't it?

> i will send a PR to remove the kubeletExtraArgs examples in https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc

:+1:

**Comment by rdxmb** [comment]
> yes, we can.

thanks!

**Review by irvifa — COMMENTED** [review]
/lgtm

**Pull request merged**
